Implement hash indexers.

sebpuetz opened the pull request:
Add implementations for FinalfusionHashIndexer and FastTextIndexer

------

The .pyi stub is autogenerated, I manually filled in the correct types. pylint doesn't pick up on the cython extensions, so I'm whitelisting failing imports from that module.

sebpuetz committed:
Implement hash indexers.

Add implementations for FinalfusionHashIndexer and FastTextIndexer

danieldk commented:
Just a heads-up, I'll review after the weekend (hopefully Monday).

danieldk commented on src/finalfusion/subword/hash_indexers.pyx:
```diff
@@ -0,0 +1,325 @@
+    @property
+    def idx_bound(self) -> int:
+        """
+        Get the **exclusive** upper bound
```
Why not call this `upper_bound` if it returns the upper bound? ;)

danieldk commented on src/finalfusion/subword/hash_indexers.pyx:
```diff
@@ -0,0 +1,325 @@
+        h = fnv64(<uint8_t*> &c, 4, h=h)
+    return h
+
+cdef uint64_t fnv64(const uint8_t*data,
```
If you hand-roll an implementation, test against the provided test vectors:

http://www.isthe.com/chongo/src/fnv/test_fnv.c

danieldk commented on src/finalfusion/subword/hash_indexers.pyx:
```diff
@@ -0,0 +1,325 @@
+        # extract unicode for char, cast to u8 pointer for hashing.
+        # extracting bytes manually from memoryview/bytes is
+        # more complex because of handling prefixes.
+        c = ord(word[i])
```
Convert to little endian to get correct hashes on big-endian systems.

danieldk commented on src/finalfusion/subword/hash_indexers.pyx:
```diff
@@ -0,0 +1,325 @@
+    cdef Py_ssize_t i = 0
+    cdef uint64_t h = SEED64
+
+    h = fnv64(<uint8_t*> &length, 8, h=h)
```
`length` must be converted to little-endian first.

danieldk commented on src/finalfusion/subword/hash_indexers.pyx:
```diff
@@ -0,0 +1,325 @@
+        if bracket:
+            word = "<%s>" % word
+        cdef bytes b_word = word.encode("utf-8")
+        cdef const unsigned char*b_word_ptr = b_word
```
space between `char` and `*`

danieldk replied:
Just another comment: `char` may not be a valid type here, depending on what `word.encode` returns. `char` is only defined by the C standard to be **at least** 1 byte (though POSIX refines it). If you want to be absolutely correct, you probably want to use `uint8_t` here in place of `unsigned char`.

danieldk commented on src/finalfusion/subword/hash_indexers.pyx:
```diff
@@ -0,0 +1,325 @@
+    def __repr__(self):
+        return f'FastTextIndexer(n_buckets={self._n_buckets}, min_n={self._min_n}, max_n={self._max_n})'
+
+cdef array.array find_utf8_boundaries(const unsigned char*w, const Py_ssize_t n_bytes):
```
space between `char` and `*`

danieldk commented on src/finalfusion/subword/hash_indexers.pyx:
```diff
@@ -0,0 +1,325 @@
+    array.resize(offsets, n_bytes + 1)
+    for b in range(n_bytes):
+        # byte w[b] is not a continuation byte, therefore beginning of char; store offset
+        if (w[b] & 0xC0) != 0x80:
```
It's really a shame that so many things have to be hand-rolled in Python/Cython :(.

danieldk replied:
Another comment here. I think this its does not lead to errors here, but note that the expression `w[b]` gets promoted to `int` with sign extension and all. It is probably nicer, more correct, and better documentation to use explicitly typed constants here for `0xc0` and `0x80`.

danieldk commented on src/finalfusion/subword/hash_indexers.pyx:
```diff
@@ -0,0 +1,325 @@
+        cdef const unsigned char*b_word_ptr = b_word
+        cdef Py_ssize_t length = len(word)
+        cdef uint32_t max_n = min(self._max_n, length)
+        cdef array.array offsets = find_utf8_boundaries(b_word_ptr, len(b_word))
```
Move up below the definition of `b_word_pointer), to separate n-gram iteration logic and UTF-8 string preparation.

danieldk reviewed (changes requested):
There are some endianness issues, which I have hopefully all marked. You should add e.g. big endian MIPS to CI to make sure that tests pass there.

You probably also want to compile the Cython code with ASAN and UBSAN and run it with those enabled for a while. I didn't immediately see any problems, but C has many undefined behaviors, so you want to be sure you are not hitting any ;).

UBSAN is also relatively cheap.

danieldk commented on src/finalfusion/subword/hash_indexers.pyx:
```diff
@@ -0,0 +1,325 @@
+cdef uint32_t PRIME32 = 16777619
+
+cdef uint64_t fifu_hash_ngram(str word, const Py_ssize_t start, const Py_ssize_t length):
+    cdef uint32_t c
```
Sorry another comment. Does Cython use C89 that all these declarations are put on top? If not and it uses >= C99, I would be in favor of declare when needed (so this would move down to where it gets assigned the code point).

sebpuetz replied:
Cython complains when I move the `cdef` into the loop, I think there was something about having to `cdef` variables in the outtermost scope of functions. So I don't think this is about C89, rather Cython's limitations.

Iteration variables also need to be defined ahead of time (although I'm not sure which of the declarations would be caught by Cython through the infer-types directive...)

sebpuetz commented:
>There are some endianness issues, which I have hopefully all marked. You should add e.g. big endian MIPS to CI to make sure that tests pass there.

I added a build on travis-ci on ARM, tests are passing for that. Setting that up on GitHub Actions seemed like a major hassle. With travis its just `arch: arm64` in `.travis.yml`

> You probably also want to compile the Cython code with ASAN and UBSAN and run it with those enabled for a while. I didn't immediately see any problems, but C has many undefined behaviors, so you want to be sure you are not hitting any ;).
>
>UBSAN is also relatively cheap.

I compiled the extension with:
```
$ CFLAGS="-fsanitize=undefined -O1" python setup.py build_ext --inplace --force --verbose --debug
$ python ngrams.py
```
and ran a script that extracted ngrams from all unique words in TWE with the FastTextIndexer and FinalfusionBucketIndexer. That didn't produce any output.

For ASAN I tried:
```
$ CFLAGS="-fsanitize=address -O1" python setup.py build_ext --inplace --force --verbose --debug
$ LD_PRELOAD=/usr/lib/gcc/x86_64-linux-gnu/7/libasan.so python asd.py
=================================================================
==13979==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 284217 byte(s) in 125 object(s) allocated from:
    #0 0x7f0e5cad5b40 in __interceptor_malloc (/usr/lib/gcc/x86_64-linux-gnu/7/libasan.so+0xdeb40)
    #1 0x557cca6092e8 in PyMem_RawMalloc Objects/obmalloc.c:520
    #2 0x557cca6092e8 in _PyObject_Malloc Objects/obmalloc.c:1579

Indirect leak of 39335 byte(s) in 41 object(s) allocated from:
    #0 0x7f0e5cad5b40 in __interceptor_malloc (/usr/lib/gcc/x86_64-linux-gnu/7/libasan.so+0xdeb40)
    #1 0x557cca6092e8 in PyMem_RawMalloc Objects/obmalloc.c:520
    #2 0x557cca6092e8 in _PyObject_Malloc Objects/obmalloc.c:1579

Indirect leak of 544 byte(s) in 1 object(s) allocated from:
    #0 0x7f0e5cad5b40 in __interceptor_malloc (/usr/lib/gcc/x86_64-linux-gnu/7/libasan.so+0xdeb40)
    #1 0x557cca609426 in PyMem_RawMalloc Objects/obmalloc.c:520
    #2 0x557cca609426 in _PyObject_Malloc Objects/obmalloc.c:1579
    #3 0x557cca609426 in pymalloc_realloc Objects/obmalloc.c:1901
    #4 0x557cca609426 in _PyObject_Realloc Objects/obmalloc.c:1920

SUMMARY: AddressSanitizer: 324096 byte(s) leaked in 167 allocation(s).
```

I then removed everything from the file, i.e. `ngrams.py` is an empty file at this point and running again yields almost the same leaks:

```
=================================================================
==14110==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 279760 byte(s) in 119 object(s) allocated from:
    #0 0x7fb905702b40 in __interceptor_malloc (/usr/lib/gcc/x86_64-linux-gnu/7/libasan.so+0xdeb40)
    #1 0x5648ad09c2e8 in PyMem_RawMalloc Objects/obmalloc.c:520
    #2 0x5648ad09c2e8 in _PyObject_Malloc Objects/obmalloc.c:1579

Indirect leak of 35879 byte(s) in 38 object(s) allocated from:
    #0 0x7fb905702b40 in __interceptor_malloc (/usr/lib/gcc/x86_64-linux-gnu/7/libasan.so+0xdeb40)
    #1 0x5648ad09c2e8 in PyMem_RawMalloc Objects/obmalloc.c:520
    #2 0x5648ad09c2e8 in _PyObject_Malloc Objects/obmalloc.c:1579

Indirect leak of 544 byte(s) in 1 object(s) allocated from:
    #0 0x7fb905702b40 in __interceptor_malloc (/usr/lib/gcc/x86_64-linux-gnu/7/libasan.so+0xdeb40)
    #1 0x5648ad09c426 in PyMem_RawMalloc Objects/obmalloc.c:520
    #2 0x5648ad09c426 in _PyObject_Malloc Objects/obmalloc.c:1579
    #3 0x5648ad09c426 in pymalloc_realloc Objects/obmalloc.c:1901
    #4 0x5648ad09c426 in _PyObject_Realloc Objects/obmalloc.c:1920

SUMMARY: AddressSanitizer: 316183 byte(s) leaked in 158 allocation(s).
```

I also spent some time on the weekend looking into valgrind with Python extensions and there's a lot of noise coming from Python itself. There are a lot of suppression files circulating. So I'm not sure whether the ASAN / leak output is really meaningful here.

--------------

This version contains the vendored fnv implementation with minimal changes (`include <stdint.h>` instead of `include "sys/types.h"`) and the `portable_endian.h`. I'm not sure whether it makes sense to distribute the fnv package as is or whether I should remove unused parts, e.g. the binaries used for testing, Makefile and README. I don't have experience in redistributing C-libs, so what is the best practice here?

danieldk commented:
> I added a build on travis-ci on ARM, tests are passing for that. Setting that up on GitHub Actions seemed like a major hassle. With travis its just `arch: arm64` in `.travis.yml`

ARM is mixed-endian, but default Linux ARM uses little-endian AFAIK.

> I also spent some time on the weekend looking into valgrind with Python extensions and there's a lot of noise coming from Python itself. There are a lot of suppression files circulating. So I'm not sure whether the ASAN / leak output is really meaningful here.

If you only compile the module with ASAN (as opposed to all of Python), the scope should be limited. Of course, Cython may still do thing that leak, I'll look in a bit more detail later.

> This version contains the vendored fnv implementation with minimal changes (`include <stdint.h>` instead of `include "sys/types.h"`) and the `portable_endian.h`.

Sounds good!

> I'm not sure whether it makes sense to distribute the fnv package as is or whether I should remove unused parts, e.g. the binaries used for testing, Makefile and README. I don't have experience in redistributing C-libs, so what is the best practice here?

I would include it as-is, maybe in a subdirectory?

sebpuetz commented:
Thanks for the comments!

> ARM is mixed-endian, but default Linux ARM uses little-endian AFAIK.

travis also supports s390x which is BE but the build fails in a lot of different places:
https://travis-ci.org/github/finalfusion/ffp/jobs/689122903

Things break with reading / writing numpy arrays for norms & storage, I'll try to fix that first before proceeding here. At least the hashing didn't break on BE...

> > I also spent some time on the weekend looking into valgrind with Python extensions and there's a lot of noise coming from Python itself. There are a lot of suppression files circulating. So I'm not sure whether the ASAN / leak output is really meaningful here.
> 
> If you only compile the module with ASAN (as opposed to all of Python), the scope should be limited. Of course, Cython may still do thing that leak, I'll look in a bit more detail later.

I only compiled the extension with ASAN / UBSAN, I installed Python through pyenv.

> > I'm not sure whether it makes sense to distribute the fnv package as is or whether I should remove unused parts, e.g. the binaries used for testing, Makefile and README. I don't have experience in redistributing C-libs, so what is the best practice here?
> 
> I would include it as-is, maybe in a subdirectory?

That's the way I have it now, it's in `src/fnv`, next to the package at `src/finalfusion`.

danieldk commented:
Could you rebase this PR on master before I review? Makes it a bit easier without having the other changes in the diff.

sebpuetz commented:
Oh sorry, I thought I had rebased on master but I rebased on the other branch, my bad.

I pushed the correct version.

Also, the travis builds failing are a bit annoying, I opened a thread in their community forums about it...

danieldk commented on setup.py:
```diff
@@ -1,19 +1,77 @@
+        from Cython.Build import cythonize
+        from Cython.Compiler import Options
+    except ImportError:
+        print("Please install Cython")
```
print to `stderr`

danieldk commented on src/fnv/longlong.h:
```diff
@@ -0,0 +1,18 @@
+#define __LONGLONG_H__
+
+/* do we have/want to use a long long type? */
+#define HAVE_64BIT_LONG_LONG	/* yes */
```
Probably safe on current platforms.

danieldk reviewed (approved):
One small comment about printing to `stderr`. Looks great otherwise.

sebpuetz commented:
Thanks! :tada:
